tests/returntypes/test_autograd_new.py

```diff
@@ -0,0 +1,1415 @@
+
+    def test_sampling(self, execute_kwargs):
+        """Test sampling works as expected"""
+        if execute_kwargs["gradient_fn"] == "device" and execute_kwargs["mode"] == "forward":
+            pytest.skip("Adjoint differentiation does not support samples")
+
+        def cost(x, device):
+            with qml.tape.QuantumTape() as tape:
+                qml.Hadamard(wires=[0])
+                qml.CNOT(wires=[0, 1])
+                qml.sample(qml.PauliZ(0))
+                qml.sample(qml.PauliX(1))
+
+            return execute([tape], device, **execute_kwargs)[0]
+
+        dev = qml.device("default.qubit", wires=2, shots=10)
+        x = np.array(0.543, requires_grad=True)
+        res = cost(x, device=dev)
+        assert isinstance(res, tuple)
+        assert len(res) == 2
+        assert res[0].shape == (10,)
+        assert res[1].shape == (10,)
+
+
+class TestHigherOrderDerivatives:
+    """Test that the autograd execute function can be differentiated"""
+
+    @pytest.mark.parametrize(
+        "params",
+        [
+            np.array([0.543, -0.654], requires_grad=True),
+            np.array([0, -0.654], requires_grad=True),
+            np.array([-2.0, 0], requires_grad=True),
+        ],
+    )
+    def test_parameter_shift_hessian(self, params, tol):
+        """Tests that the output of the parameter-shift transform
+        can be differentiated using autograd, yielding second derivatives."""
+        dev = qml.device("default.qubit.autograd", wires=2)
+        params = np.array([0.543, -0.654], requires_grad=True)
+
+        def cost_fn(x):
+            with qml.tape.QuantumTape() as tape1:
+                qml.RX(x[0], wires=[0])
+                qml.RY(x[1], wires=[1])
+                qml.CNOT(wires=[0, 1])
+                qml.var(qml.PauliZ(0) @ qml.PauliX(1))
+
+            with qml.tape.QuantumTape() as tape2:
+                qml.RX(x[0], wires=0)
+                qml.RY(x[0], wires=1)
+                qml.CNOT(wires=[0, 1])
+                qml.probs(wires=1)
+
+            result = execute([tape1, tape2], dev, gradient_fn=param_shift, max_diff=2)
+            return result[0] + result[1][0]
+
+        res = cost_fn(params)
+        x, y = params
+        expected = 0.5 * (3 + np.cos(x) ** 2 * np.cos(2 * y))
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = qml.grad(cost_fn)(params)
+        expected = np.array(
+            [-np.cos(x) * np.cos(2 * y) * np.sin(x), -np.cos(x) ** 2 * np.sin(2 * y)]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = qml.jacobian(qml.grad(cost_fn))(params)
+        expected = np.array(
+            [
+                [-np.cos(2 * x) * np.cos(2 * y), np.sin(2 * x) * np.sin(2 * y)],
+                [np.sin(2 * x) * np.sin(2 * y), -2 * np.cos(x) ** 2 * np.cos(2 * y)],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_adjoint_hessian_one_param(self, tol):
```

Comment:
There seem to be two gate parameters that are trainable :thinking: I think it's worth commenting that only one may affect the gradient and hence the Hessian.

Reply:
Yes but Autograd consider that as one parameter, that's why you have only an array as return.